File: src/main/java/me/narutopig/bors/lifesteal/listeners/Lifesteal.java
package me.narutopig.bors.lifesteal.listeners;

import org.bukkit.attribute.Attribute;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class Lifesteal implements Listener {
    private final Map<UUID, Long> killManager = new HashMap<>(); // keeps tracked of how long ago you were killed

    public static void setMaxHealth(Player player, double hearts) {
        player.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(hearts);
    }

    @EventHandler
    public void onPlayerDeath(PlayerDeathEvent event) {
        Player killed = event.getEntity();
        Player killer = killed.getKiller();

        if (killer == null) return;

        UUID uuid = killed.getUniqueId();

        long currTime = System.currentTimeMillis();
        if (currTime - killManager.getOrDefault(uuid, 0L) >= 15L * 1000) {
            // runs only if the last time the victim was killed was more than 15 seconds ago
            setMaxHealth(killed, killed.getMaxHealth() - 2);
            setMaxHealth(killed, killer.getMaxHealth() + 2);
        }
        
        killManager.put(uuid, currTime);
    }
}

File: src/main/kotlin/io/github/narutopig/bors/listeners/Lifesteal.kt
package io.github.narutopig.bors.listeners

import io.github.narutopig.bors.Main
import org.bukkit.Material
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerInteractEvent

class Lifesteal : Listener {
    @EventHandler
    fun onPlayerInteract(event: PlayerInteractEvent) {
        if (!Main.configuration.getBoolean("plugins.lifesteal")) return

        if (event.action == Action.RIGHT_CLICK_AIR) {
            val hand = event.player.inventory.itemInMainHand
            if (hand.type == Material.NETHER_STAR) {
                if (hand.amount > 0 && event.player.maxHealth <= 38) {
                    event.player.inventory.itemInMainHand.amount -= 1
                    event.player.maxHealth += 2
                }
            }
        }
    }

    @EventHandler
    fun onPlayerDeath(event: PlayerDeathEvent) {
        if (!Main.configuration.getBoolean("plugins.lifesteal")) return

        val deadPlayer = event.entity
        val killer = deadPlayer.killer
        if (killer == null) {
            val hp = deadPlayer.maxHealth
            if (hp >= 4) {
                event.entity.maxHealth = hp - 2
            }
        } else {
            if (deadPlayer.maxHealth >= 4 && killer.maxHealth <= 38) { // hearts change only if killed player has at least two hearts and killer has nineteen hearts
                val hp = deadPlayer.maxHealth
                event.entity.maxHealth = hp - 2
                if (event.entity.killer != null) {
                    val hp2 = killer.maxHealth
                    event.entity.killer!!.maxHealth = hp2 + 2
                }
            }
        }
    }
}